### ISSUE TITLE
[RFC]: Runtime Weight Synchronization HTTP API for Qwen3-Image

### ISSUE THREAD

**Issue by princepride** [vllm-project/vllm-omni#6265]
## Motivation.

Online RL/fine-tuning workflows need to refresh rollout workers with newly trained
Qwen3-Image weights without restarting the vLLM-Omni server. Today vLLM-Omni can
serve the Qwen-Image family and its pipelines expose `load_weights()`, while the
engine already provides a `collective_rpc` path to diffusion workers. However,
there is no supported HTTP contract that coordinates a live weight update.

Without such a contract, integrations must restart the server or call internal
Python methods. Restarts are slow and lose warm runtime state; direct internal
calls do not define request draining, multi-rank consistency, versioning, cache
invalidation, authentication, or failure behavior. A partially updated replica
can also serve an image using weights from different training steps.

This RFC proposes an opt-in administrative HTTP API for atomic, versioned weight
synchronization, initially scoped to the Qwen3-Image diffusion transformer (DiT).

### Goals

1. Update Qwen3-Image DiT weights without restarting the serving process.
2. Ensure every worker/rank in the selected replica switches to the same version.
3. Never mix old and new weights within one generation request.
4. Return machine-readable progress and errors to training/rollout controllers.
5. Reuse the existing engine, stage, and diffusion-worker RPC hierarchy.
6. Keep the API disabled by default and suitable for trusted administrative use.

### Non-goals for P0

- Sending multi-gigabyte tensor payloads directly in the HTTP request.
- Updating the text encoder, tokenizer, VAE, or scheduler.
- Hot-swapping to a checkpoint with a different model architecture or parameter
  shape.
- Supporting every diffusion pipeline in the first implementation.
- Guaranteeing rollback after an in-place CUDA write has started. A failed
  update makes the replica unavailable until it is recovered or restarted.

### Public API

Enable the route explicitly:

```bash
vllm serve Qwen/Qwen3-Image --omni \
  --enable-weight-sync-api \
  --weight-sync-root /checkpoints
```

Start a synchronization:

```http
POST /v1/weights/sync
Authorization: Bearer <api-key>
Content-Type: application/json

{
  "model": "Qwen/Qwen3-Image",
  "component": "transformer",
  "source": {
    "type": "local_checkpoint",
    "path": "/checkpoints/step-1200/transformer"
  },
  "version": "step-1200",
  "sha256_manifest": "9c1c...",
  "wait": true,
  "timeout_seconds": 600
}
```

Successful synchronous response:

```json
{
  "id": "ws_01J...",
  "status": "succeeded",
  "model": "Qwen/Qwen3-Image",
  "component": "transformer",
  "previous_version": "step-1100",
  "version": "step-1200",
  "updated_workers": 8,
  "duration_ms": 18342
}
```

When `wait=false`, return HTTP 202 with `status="queued"` or `status="running"`.
Clients can inspect the operation and active version:

```http
GET /v1/weights/sync/{id}
GET /v1/weights/version
```

P0 accepts only absolute local/shared-filesystem paths resolved under
`--weight-sync-root`. Remote URLs are rejected to avoid SSRF and inconsistent
downloads across workers. Object-store and NCCL/tensor-stream transports can be
added later behind new `source.type` values without changing the control-plane
contract.

### Status and error contract

| HTTP status | Meaning |
| --- | --- |
| 200 | Synchronous update completed, or idempotent version already active. |
| 202 | Asynchronous update accepted. |
| 400 | Malformed request or unsupported component. |
| 401/403 | Administrative API authentication/authorization failed. |
| 404 | Checkpoint, model, or operation not found. |
| 409 | Another update is active, version conflict, or requests cannot drain. |
| 422 | Manifest, tensor names, shapes, dtypes, or architecture are incompatible. |
| 500 | Update failed before a safe commit. |
| 503 | One or more ranks may be inconsistent; replica is fenced from inference. |

Errors use the existing OpenAI-compatible error envelope. Repeating a request
with the same `version` and manifest is idempotent. Reusing a version with a
different manifest returns 409.

### Update lifecycle

```mermaid
sequenceDiagram
    participant C as Trainer / Controller
    participant H as HTTP API
    participant O as AsyncOmni / Orchestrator
    participant S as Diffusion Scheduler
    participant W as Qwen3-Image Workers

    C->>H: POST /v1/weights/sync
    H->>H: Authenticate and validate path/version
    H->>O: Begin update transaction
    O->>S: Pause new admissions for target replica
    S-->>O: In-flight requests drained
    O->>W: collective_rpc(prepare_weight_update)
    W-->>O: Manifest/name/shape validation result
    O->>W: collective_rpc(load_weight_update)
    W->>W: Load transformer tensors and synchronize device
    W-->>O: Per-rank digest and loaded-name summary
    O->>W: collective_rpc(commit_weight_update)
    W->>W: Invalidate compiled graphs and weight-dependent caches
    O->>S: Resume admissions with new version
    O-->>H: Committed version on every rank
    H-->>C: 200 / 202 result
```

The scheduler must stop admitting requests to the target replica and wait for
already admitted requests to finish before any tensor is mutated. A configurable
drain timeout prevents the admin request from waiting forever. This is a
replica-local barrier; other replicas may continue serving.

### Worker implementation

Add a model-agnostic worker protocol with a Qwen3-Image implementation:

```python
class SupportsWeightSync(Protocol):
    def prepare_weight_update(self, spec: WeightUpdateSpec) -> WeightManifest: ...
    def load_weight_update(self, spec: WeightUpdateSpec) -> WeightLoadResult: ...
    def commit_weight_update(self, version: str) -> None: ...
```

The HTTP handler sends the operation through `AsyncOmniEngine.collective_rpc`,
the orchestrator selects the Qwen3-Image diffusion stage/replica, and the
diffusion executor invokes the method on every worker. The pipeline delegates
tensor assignment to the existing Qwen-Image `load_weights()` implementation.

P0 checkpoint rules:

- safetensors only;
- transformer keys only, with an optional `transformer.` prefix;
- exact parameter-name and shape match;
- explicit policy for dtype conversion (default: reject mismatches);
- all expected trainable transformer parameters must be present;
- unexpected keys and partial updates are rejected;
- every rank reports the same manifest digest and resulting version.

For tensor/sequence parallel deployments, each worker reads the checkpoint and
uses the model's existing sharded weight loader. Loading must happen on the
diffusion engine's serialized RPC path so it cannot race normal model execution.

### Atomicity and failure fencing

The externally visible commit point occurs only after every rank successfully
loads and verifies the requested version. No new inference request can enter the
replica between drain and commit.

P0 uses in-place loading and therefore cannot promise rollback after mutation.
If any rank fails after loading begins:

1. keep admission paused;
2. mark the operation `failed_inconsistent`;
3. mark the replica unhealthy and return HTTP 503;
4. require an explicit retry with a complete checkpoint or a process restart.

Future work may provide double-buffered modules and pointer-swap commit where
memory capacity allows true rollback.

### Cache and execution-state invalidation

After all ranks load successfully and before admission resumes:

- clear TeaCache, Cache-DiT, and other weight-dependent caches;
- invalidate or recapture CUDA graphs / `torch.compile` artifacts whose guards
  do not safely permit parameter replacement;
- synchronize CUDA/NPU devices;
- preserve tokenizer, text-encoder, VAE, and scheduler state because P0 does not
  update them;
- expose the committed `weight_version` in logs, metrics, and response metadata.

Inference responses should include `X-vLLM-Omni-Weight-Version`. This makes
rollout samples traceable to the training checkpoint that produced them.

### Concurrency and replica behavior

- One active weight update per replica.
- Duplicate requests for the same version/manifest join or return the existing
  operation.
- Different versions conflict with an active update.
- A deployment controller updates replicas one at a time for availability, then
  verifies `GET /v1/weights/version` on every replica.
- P0 does not claim cluster-wide atomicity across independent API-server
  processes; that belongs to the external controller.

### Security

- Route disabled unless `--enable-weight-sync-api` is set.
- Reuse configured API-key authentication and document that the endpoint must
  not be exposed publicly.
- Restrict checkpoint paths to canonical paths beneath `--weight-sync-root`;
  reject traversal and symlink escapes.
- Do not accept pickle checkpoints or arbitrary Python loader hooks.
- Log requester identity (when available), source, version, manifest, result,
  and timing without logging secrets.

### Observability

Add metrics:

- `vllm_omni_weight_sync_requests_total{status}`
- `vllm_omni_weight_sync_duration_seconds`
- `vllm_omni_weight_sync_active`
- `vllm_omni_weight_version_info{model,component,version}`
- `vllm_omni_weight_sync_failures_total{phase}`

#### P0: Qwen3-Image shared-filesystem synchronization

- API schemas, opt-in route, authentication, and operation status store.
- Replica drain/fence integration.
- Collective RPC prepare/load/commit methods.
- Qwen3-Image transformer safetensors loader.
- Version header, metrics, cache invalidation, and failure fencing.

#### P1: Training-system integration and faster transports

- Object-store manifests with per-worker staged downloads.
- NCCL or tensor-stream data plane for colocated trainers and rollout workers.
- Rolling multi-replica coordinator.
- Optional partial/delta updates with an explicit compatibility contract.

#### P2: Generalization

- Promote the protocol to other validated diffusion pipelines and components.
- Double-buffered atomic swap where memory permits.

#### Unit tests

- Request validation, authentication, path containment, idempotency, and status
  transitions.
- Manifest validation for missing, unexpected, mismatched-shape, and mismatched-
  dtype tensors.
- Admission pause, in-flight drain, timeout, resume, and concurrent-update
  conflict behavior.
- All-rank success, one-rank prepare failure, one-rank load failure, and replica
  fencing.
- Cache/compile invalidation and committed version propagation.

#### Distributed integration tests

- Tiny Qwen3-Image checkpoint on 1 GPU and TP=2.
- Generate with version A, synchronize to deliberately different version B,
  and verify output changes for the same prompt/seed.
- Verify requests admitted before the barrier complete entirely on A and those
  admitted after commit run entirely on B.
- Verify all workers report the same version and manifest digest.
- Inject a rank failure and verify the replica serves no mixed-version output.

#### End-to-end rollout test

A training/controller process writes step N safetensors, calls the HTTP API,
waits for success, confirms the version endpoint, and only then begins the next
rollout batch. Each returned sample records the expected weight version.

### Acceptance criteria

1. Qwen3-Image transformer weights can be updated through HTTP without process
   restart on supported single-GPU and tensor-parallel configurations.
2. No request observes a mixture of two weight versions.
3. A successful response means all target ranks committed the requested version.
4. Any post-mutation rank failure fences the replica instead of serving traffic.
5. Repeating the same update is idempotent.
6. Requests and outputs are traceable to a stable weight version.
7. The feature is disabled by default and rejects unsafe checkpoint sources.

### Relationship to existing work

- Reuse the current diffusion `collective_rpc` serialization/routing path.
- Reuse Qwen-Image pipeline and transformer `load_weights()` implementations.
- Complement deterministic rollout RFC #4864: determinism controls how a fixed
  version is sampled, while this RFC defines when a new version becomes visible.
- Coordinate route ownership with API-server refactor RFC #5227.

### Open questions

1. Is shared-filesystem safetensors sufficient for P0, or is an NCCL data plane
   required by the first training integration?
2. Should the route be `/v1/weights/sync` or an explicitly administrative path
   such as `/v1/admin/weights/sync`?
3. Should P0 expose only synchronous behavior, or retain the proposed async
   operation resource from the start?
4. Can current compile/CUDA-graph modes safely observe in-place parameter writes,
   or must P0 reject/disable those modes during live updates?
5. Which exact Qwen3-Image checkpoint naming convention should be canonicalized
   at the HTTP boundary?

## Feedback Period.

Two weeks from publication. The API shape, failure semantics, and first required
transport should be agreed before implementation begins.

## CC List.

Maintainers of Qwen-Image, diffusion serving, distributed executors, API server,
and verl-omni integration are invited to review.

## Any Other Things.

- The initial implementation should remain behind an experimental flag until
  distributed failure-injection tests pass.
- This RFC was prepared after searching existing vLLM-Omni RFCs and repository
  documentation; no existing Qwen3-Image HTTP weight-synchronization RFC was
  found.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot
  living at the bottom right corner of the documentation page, which can answer
  lots of frequently asked questions.
